// Generated by https://github.com/foxglove/schemas

syntax = "proto3";

import "google/protobuf/timestamp.proto";

package foxglove;

// FOXGLOVE: A navigation satellite fix for any Global Navigation Satellite System
message LocationFix {

  // Timestamp of the message
  google.protobuf.Timestamp timestamp = 6;

  // Frame for the sensor. Latitude and longitude readings are at the origin of the frame.
  string frame_id = 7;

  // Latitude in degrees
  double latitude = 1;

  // Longitude in degrees
  double longitude = 2;

  // Altitude in meters
  double altitude = 3;

  // Position covariance (m^2) defined relative to a tangential plane through the reported position. The components are East, North, and Up (ENU), in row-major order.
  repeated double position_covariance = 4; // length 9

    // Type of position covariance
    enum PositionCovarianceType {
      UNKNOWN = 0;
      APPROXIMATED = 1;
      DIAGONAL_KNOWN = 2;
      KNOWN = 3;
    }
  

  // If `position_covariance` is available, `position_covariance_type` must be set to indicate the type of covariance.
  PositionCovarianceType position_covariance_type = 5;
}


// Keeloson Recommended PositionFix message
message PositionFix {

  // Timestamp of the message
  google.protobuf.Timestamp timestamp = 1;

  // Latitude in degrees
  double latitude_degrees = 3;

  // Longitude in degrees
  double longitude_degrees = 4;

  // Altitude in meters
  double altitude_meters = 5;

  // Horizontal accuracy in meters 
  double horizontal_accuracy_meters = 6;

  // Vertical accuracy in meters
  double vertical_accuracy_meters = 7;

  // Accuracy of the position fix in meters
  double accuracy_meters = 8;

  // Geodetic datum used for the position fix (e.g. WGS84)
  string geodetic_datum = 9;

}

message PositionSourceSatellites {

  // Timestamp of the message
  google.protobuf.Timestamp timestamp = 1;

  // The source of the position fix
  enum Source {
    UNKNOWN = 0;
    GPS = 1;
    GLONASS = 2;
    GALILEO = 3;
    BEIDOU = 4;
    SBAS = 5;
    QZSS = 6;
    IRNSS = 7;
    NAVIC = 8;
    OTHER = 9;
  }

  // The source of the position fix
  repeated Source source = 2;

  enum CorrectionSource {
    NONE = 0;
    DGPS = 1;
    RTK = 2;
  }
  CorrectionSource correction_source = 3;

  // Satellite used for the position fix
  int32 satellites_used = 4;

  // Dilution of Precision: indicates the quality of the GPS signal. Lower values represent better accuracy
  // A value of 1 or bellow indicates a ideal signal, while a value over 20 indicates a very poor signal.

  // DOP Type  | Stands for                        | What it measures
  // ----------|-----------------------------------|-------------------------------
  //    HDOP   | Horizontal Dilution of Precision  | Latitude + Longitude accuracy
  //    VDOP   | Vertical Dilution of Precision    | Altitude (vertical) accuracy
  //    PDOP   | Position Dilution of Precision    | 3D position accuracy

  // HDOP - Horizontal Dilution of Precision
  double hdop = 5;
  
  // VDOP - Vertical Dilution of Precision
  double vdop = 6;
  
  // PDOP Position Dilution of Precision
  double pdop = 7;
}



